### ISSUE TITLE
Probably `selectedId` is better than `selectedIndex` in ComboBox and Dropdown

### ISSUE THREAD

**Issue by KoichiKiyokawa** [carbon-design-system/carbon-components-svelte#1003]
First, thank you for working on this library.
I use this library in production and it is very useful 🎉 
But, I found that the reactive prop of `selectedIndex` in `ComboBox` and `Dropdown` was difficult to use.
It would be better if it is `selectedId`.

The reasons are as follows.
## Matches with other select like components.
- `Select` component has a `selected` prop (it is equivalent to selectedId).
- `MultiSelect` component has a `selectedIds` prop.

## Easier to connect with server side
When sending values to the server side, it is often necessary to use the `selectedId` instead of the `selectedIndex`.

## Matches HTML `<select>` tags.
Since `ComboBox` and `Dropdown` are used as alternatives to the `<select>` tag, it is better if the APIs match, IMO.

```svelte
<!-- select tag use case in svelte-->
<script>
let selectedFruit
</script>

<select bind:value={selectedFruit}>
  <option value="apple">Apple</option>
  <option value="banana">Banana</option>
</select>
```

Currently, it is necessary to declare an extra variable when using `ComboBox` or `Dropdown`.

```svelte
<script>
const items = [{ id: 'apple', text: 'Apple' }, { id: 'banana', text: 'Banana' }]
let selectedFruitIndex = -1 // the extra variable
$: selectedFruit = items[selectedIndex] // we need this to send the server side.
</script>

<Dropdown {items} bind:selectedIndex={selectedFruitIndex} />
```

or, to avoid extraneous variables, do the following.

```svelte
<script>
const items = [{ id: 'apple', text: 'Apple' }, { id: 'banana', text: 'Banana' }]
let selectedFruit
</script>

<Dropdown {items} selectedIndex={items.findIndex(item => item.id === selectedFruit)} on:select={e => selectedFruit = e.detail.selectedId} />
```

If `ComboBox` and `Dropdown` have a reactive `selectedId`, we can do the following. It is very smart.

```svelte
<script>
let selectedFruit
</script>

<Dropdown items={[{ id: 'apple', text: 'Apple' }, { id: 'banana', text: 'Banana' }]} bind:selectedId={selectedFruit} />
```

This issue is just a suggestion, so feel free to close 👍

**Comment by metonym** [comment]
I agree.

It would be nice to be consistent and use `selectedId`. Setting `selectedId` to `undefined` would clear. However, one thing to consider is that `Dropdown` must have a selected value at all times.

**Comment by KoichiKiyokawa** [comment]
@metonym 
Thank you for your reply.

> Dropdown must have a selected value at all times.

Currently, the default value of `selectedIndex` is -1 in `Dropdown`.
It means `Dropdown` can be unselected, isn't it?

When unselected, the `label` prop will be shown.
https://github.com/carbon-design-system/carbon-components-svelte/blob/32edfa59dddbb1346b6564192a6bf130e80602a9/src/Dropdown/Dropdown.svelte#L227

**Comment by metonym** [comment]
@KoichiKiyokawa 👍 Are you interested in working on this?

**Comment by KoichiKiyokawa** [comment]
@metonym 
Yes, I will try!

**Comment by metonym** [comment]
Making a note that this would resolve #886

**Comment by patoi** [comment]
I have previously suggested using `selectId` in a PR: https://github.com/carbon-design-system/carbon-components-svelte/pull/926#issuecomment-1005790133

We created an extension of ComboBox in our projects, that it brings model based working.

How it works:

```
<ComboBoxExtension
        bind:selectId
        list
        listMap={element => ({ id: element.aKey, text: element.aValue })}
        ...
```
`listMap` is optional, mapping an arbitrary object to `{id: string, text: string}`
The `selectedIndex` variable has been hiden.

**Comment by patoi** [comment]
Dropdown could be a variant of the ComboBox, where:

- inputValue is not editable
- there is no filter (no inputValue)
- could be empty as the ComboBox (undefined, null will be an empty string, and a placeholder will appear as in the TextInput)

**Comment by alkanna** [comment]
Just to be clear, this does not induce a breaking change because we would not remove the `selectedIndex` prop ?

**Comment by metonym** [comment]
@alkanna I believe it would break two-way binding on `selectedIndex`.

For example, you would still be able to programmatically update `selectedIndex`, but the initial value (1) would not work (the initial value would be `undefined`).

```svelte
<script>
  import { Dropdown } from "carbon-components-svelte";

  let selectedIndex = 1;

  const items = [
    { id: "0", text: "Slack" },
    { id: "1", text: "Email" },
    { id: "2", text: "Fax" },
  ];
</script>

<Dropdown titleText="Contact" bind:selectedIndex {items} />
```

**Comment by metonym** [comment]
However, `selectedIndex` will be replaced entirely by `selectedId` in @KoichiKiyokawa's PR

**Comment by KoichiKiyokawa** [comment]
@alkanna
Because selectedId and selectedIndex are interdependent, it is difficult to make both of the following work.
- When `selectedIndex` is programmatically updated, the component updates `selectedId`
- When `selectedId` is programmatically updated, the component updates `selectedIndex`

I tried to avoid infinite loop, but there was a compile error.
```
Cyclical dependency detected: selectedIndex → selectedId → selectedIndex
```
https://svelte.dev/repl/ec44a887a7a3433888566283c43ce9a2?version=3.46.2

There may be a way to make both selectedId and selectedIndex compatible, but I think that would be tricky and likely to increase maintenance costs.
IMO, `selectedIndex` is not as useful as the `selectedId`, so it might be a good idea to delete it while the major version is 0.

**Comment by alkanna** [comment]
@KoichiKiyokawa Thank you for the detailed explanation. I am pretty sure everyone here using this library is doing some gymnastics to get any dynamically generated ComboBox or Dropdown to work properly because of this. This change would make two way databinding a breeze.

**Comment by metonym** [comment]
Released in [v0.53.0](https://github.com/carbon-design-system/carbon-components-svelte/releases/tag/v0.53.0).
